lib/base/atomic-file.cpp

```diff
@@ -0,0 +1,92 @@
+/* Icinga 2 | (c) 2022 Icinga GmbH | GPLv2+ */
+
+#include "base/atomic-file.hpp"
+#include "base/exception.hpp"
+#include "base/utility.hpp"
+#include <utility>
+
+#ifdef _WIN32
+#	include <io.h>
+#	include <windows.h>
+#else /* _WIN32 */
+#	include <errno.h>
+#	include <unistd.h>
+#endif /* _WIN32 */
+
+using namespace icinga;
+
+AtomicFile::AtomicFile(String path, int mode) : m_Path(std::move(path))
+{
+	m_TempFilename = m_Path + ".tmp.XXXXXX";
+	int fd;
+
+#ifdef _WIN32
+	fd = Utility::MksTemp(&m_TempFilename[0]);
+#else /* _WIN32 */
+	fd = mkstemp(&m_TempFilename[0]);
+#endif /* _WIN32 */
+
+	if (fd < 0) {
+		auto error (errno);
+
+		BOOST_THROW_EXCEPTION(posix_error()
+			<< boost::errinfo_api_function("mkstemp")
+			<< boost::errinfo_errno(error)
+			<< boost::errinfo_file_name(m_TempFilename));
+	}
+
+	exceptions(failbit | badbit);
+	open(boost::iostreams::file_descriptor(fd, boost::iostreams::close_handle));
```

Comment:
Due to being surprised that this line works at all (wasn't expecting that you could construct `file_descriptor` from an `int` rather than from a `HANDLE`), I had a closer look and it turned out this has an issue with leaking fds: https://github.com/boostorg/iostreams/issues/152

So I guess for now, on Windows opening the file by name seems to be the other option. For using `HANDLE` we'd have to replace that `Utility::MksTemp` code :(